[MRG] Unit/dimensions refactoring

mstimberg opened the pull request:
This implements what I proposed in #797: `SingleEquation`, `Variable`, and `VariableView` now internally store dimensions instead of units. If I run a HH model repeatedly in a loop (with `run(0*ms)`) to coarsely matter preparation overhead, then this actually cuts the time in half! This mostly shows how inefficient `get_unit` was (which again and again went through the full registry to search for a suitable unit, checking and discarding scaled units along the way), but still. I also profiled a bit where the rest of the time is spend, and I think we can quite easily improve more with some simple caching. I might tackle this soon.
Note that you can still access a `unit` property of the above mentioned classes (since it was kind of part of our "public API" and I have already seen user code making use of this, e.g. to create equation strings programmatically), but this will raise a deprecation warning.

Fixes #797

thesamovar commented:
Haven't looked at the code yet, but two possibilites:

1. Maybe we ought to improve ``get_unit`` even if we're not using it here if it's that inefficient? We could probably, given a quantity ``x`` construct the best unit explicitly via a lookup table of ``x.dimension`` followed by getting the best scale by inspecting the value of ``int(log10(float(x)))``? I guess this might not work with user constructed units quite as nicely?
2. I'd be happy with leaving a computed ``unit`` attribute and not having the ``DeprecationWarning`` - unless there's a problem with that?

mstimberg commented:
> Maybe we ought to improve get_unit even if we're not using it here if it's that inefficient? We could probably, given a quantity x construct the best unit explicitly via a lookup table of x.dimension followed by getting the best scale by inspecting the value of int(log10(float(x)))? I guess this might not work with user constructed units quite as nicely?

Actually, `get_unit` is only used to get the unscaled unit for dimensions, e.g. when we want to display `volt` instead of `m^2 kg s^-3 A^-1` in an error message. To get the "best unit" for a quantity, we use `Quantity.get_best_unit` (which I just made "public", it was called `_get_best_unit` before). The problem with `get_unit` was that the algorithm was the following:
```Python
for u in all_registered_units(*regs):
    if np.array(u, copy=False) == 1 and have_same_dimensions(u, x):
        return u
```
So it went through the whole list of units every time, making a lot of checks before finding the appropriate unit... A simple caching solution was not feasible given the flexibility to pass in arbitrary lists of `UnitRegistry` objects (the `*regs` argument). Well, we never actually use that (I can't really think of a use-case: we have a `UnitRegistry` for user-defined units, not sure what else we could need), so I got rid of this flexibility and also made use of the already existing dictionary that maps dimensions to units. Et voilà, `get_unit` is now about 500-4000 times faster :-) So I think we can close this chapter now. I made a few other minor related changes (e.g. removing the `all_registered_units` generator which was not used anywhere anymore) and also changed the order of lookup (now the user-defined units are first instead of last).

> I'd be happy with leaving a computed unit attribute and not having the DeprecationWarning - unless there's a problem with that?

That does make sense, I removed it. Given that `get_unit` is now fast it is not bad to use it anyway -- in our internal code we should simply take care to not use units where dimensions suffice.

That reminds me of something: I did actually not change units into dimensions in *all* places where it makes sense: the `check_units` decorator still uses units even though it strictly checks for correct dimensions. I think users don't generally care about this distinction and talking about "units" is more natural then talking about "dimensions" (also writing e.g. `volt` instead of `volt.dim`...).

thesamovar commented:
Agreed on continuing to use units for user-facing code rather than dimensions! Will take a look at the code asap, but sounds good to go.

thesamovar commented on brian2/core/variables.py:
```diff
@@ -296,17 +291,17 @@ class Constant(Variable):
-        The unit of the variable. Note that the variable itself (as referenced
-        by value) should never have units attached.
+    dimensions : `Dimension`, optional
+        The dimensions of the variable. Note that the variable itself (as
```
Again. I know that it just said dimensions before, but seems like a good opportunity to clean up potential source of confusion.

thesamovar commented on brian2/core/variables.py:
```diff
@@ -356,17 +351,17 @@ class AuxiliaryVariable(Variable):
-    unit : `Unit`
-        The unit of the variable.
+    dimensions : `Dimension`, optional
+        The dimensions of the variable.
```
And here.

thesamovar commented on brian2/core/variables.py:
```diff
@@ -387,8 +382,8 @@ class ArrayVariable(Variable):
-    unit : `Unit`
-        The unit of the variable
+    dimensions : `Dimension`, optional
+        The dimensions of the variable
```
Apparently we say this a lot.

thesamovar commented on brian2/core/variables.py:
```diff
@@ -483,8 +479,8 @@ class DynamicArrayVariable(ArrayVariable):
-    unit : `Unit`
-        The unit of the variable
+    dimensions : `Dimension`, optional
+        The dimensinos of the variable
```
And again, but this time there is even a typo to correct! ;)

thesamovar commented on brian2/core/variables.py:
```diff
@@ -519,18 +515,18 @@ class DynamicArrayVariable(ArrayVariable):
 
         #: The number of dimensions
-        self.dimensions = dimensions
+        self.ndim = ndim
```
Yeah I like using ndim instead of dimensions, that's much better!

thesamovar commented on brian2/core/variables.py:
```diff
@@ -583,8 +578,8 @@ class Subexpression(Variable):
-    unit : `Unit`
-        The unit of the subexpression.
+    dimensions : `Dimension`, optional
+        The dimensions of the subexpression.
```
!

thesamovar commented on brian2/core/variables.py:
```diff
@@ -734,11 +730,11 @@ class VariableView(object):
-        The unit to be used for the variable, should be `None` when a variable
-         is accessed without units (e.g. when accessing ``G.var_``).
+    dimensions : `Dimension`, optional
+        The dimensinos to be used for the variable, should be `None` when a
```
And again with a typo.

thesamovar commented on brian2/core/variables.py:
```diff
@@ -761,14 +757,14 @@ def __init__(self, name, variable, group, unit=None):
         # indexing is still possible, even if the group no longer exists
         self.indexing = self.group._indices
-        self.unit = unit
+        self.dim = dimensions
```
I'm actually slightly concerned that we have dim and ndim now that I think about it, but I think it's probably OK.

thesamovar commented on brian2/core/variables.py:
```diff
@@ -1418,18 +1413,18 @@ def _add_variable(self, name, var, index=None):
-        unit : `Unit`
-            The unit of the variable
+        dimensions : `Dimension`, optional
+            The dimensions of the variable.
```
!

thesamovar commented on brian2/core/variables.py:
```diff
@@ -1491,8 +1486,8 @@ def add_arrays(self, names, unit, size, values=None, dtype=None,
-        unit : `Unit`
-            The unit of the variables
+        dimensions : `Dimension`, optional
+            The dimensions of the variable.
```
!

thesamovar commented on brian2/core/variables.py:
```diff
@@ -1530,8 +1526,8 @@ def add_dynamic_array(self, name, unit, size, values=None, dtype=None,
-        unit : `Unit`
-            The unit of the variable
+        dimensions : `Dimension`, optional
+            The dimensions of the variable.
```
!

thesamovar commented on brian2/core/variables.py:
```diff
@@ -1609,40 +1605,40 @@ def add_arange(self, name, size, start=0, dtype=np.int32, constant=True,
         value: reference to the variable value
             The value of the constant.
+        dimensions : `Dimension`, optional
+            The dimensions of the variable. Note that the variable itself (as
```
!

thesamovar commented on brian2/equations/equations.py:
```diff
@@ -203,7 +201,7 @@ def check_identifier_constants(identifier):
-def unit_and_type_from_string(unit_string):
+def dimensions_and_type_from_string(unit_string):
     '''
     Returns the unit that results from evaluating a string like
```
I guess this should now read physical dimensions instead of unit.

thesamovar commented on brian2/equations/equations.py:
```diff
@@ -217,8 +215,8 @@ def unit_and_type_from_string(unit_string):
-    u, type : (Unit, {FLOAT, INTEGER or BOOL})
-        The resulting unit and the type of the variable.
+    d, type : (`Dimension`, {FLOAT, INTEGER or BOOL})
+        The resulting dimensions and the type of the variable.
```
!

thesamovar commented on brian2/equations/equations.py:
```diff
@@ -343,8 +341,8 @@ class SingleEquation(object):
-    unit : Unit
-        The unit of the variable
+    dimensions : `Dimension`
+        The dimensions of the variable
```
!

thesamovar commented on brian2/equations/equations.py:
```diff
@@ -767,9 +767,9 @@ def _get_stochastic_type(self):
+    dimensions = property(lambda self: dict([(var, eq.dim) for var, eq in
                                        self._equations.iteritems()]),
-                     doc='Dictionary of all internal variables and their corresponding units.')
+                     doc='Dictionary of all internal variables and their corresponding dimensions.')
```
!

thesamovar commented on brian2/equations/unitcheck.py:
```diff
@@ -23,8 +24,8 @@ def check_unit(expression, unit, variables):
-    unit : `Unit`
-        The expected unit for the `expression`.
+    dimensions : `Dimension`
+        The expected dimensions for the `expression`.
```
!

thesamovar commented on brian2/equations/unitcheck.py:
```diff
@@ -36,10 +37,12 @@ def check_unit(expression, unit, variables):
-                                                  'have the expected unit %r') %
-                                                  (expression.strip(), unit))
+    expr_dims = parse_expression_dimensions(expression, variables)
+    # FIXME: The error message is not great
```
Will fix before merging?

thesamovar commented on brian2/groups/neurongroup.py:
```diff
@@ -111,10 +111,10 @@ def _get_refractory_code(self, run_namespace):
-            unit = parse_expression_unit(str(ref), variables)
-            if have_same_dimensions(unit, second):
+            dims = parse_expression_dimensions(str(ref), variables)
+            if dims is second.dim:
```
Same as before. For example, might we not have something that evaluates as e.g. ``sqrt(second*second)`` which would fail ``is second.dim``?

thesamovar commented on brian2/groups/neurongroup.py:
```diff
@@ -111,10 +111,10 @@ def _get_refractory_code(self, run_namespace):
+            if dims is second.dim:
                 abstract_code = 'not_refractory = (t - lastspike) > %s\n' % ref
-            elif have_same_dimensions(unit, Unit(1)):
+            elif dims is DIMENSIONLESS:
```
Same issue?

thesamovar commented on brian2/groups/neurongroup.py:
```diff
@@ -128,7 +128,7 @@ def _get_refractory_code(self, run_namespace):
                 raise TypeError(('Refractory expression has to evaluate to a '
                                  'timespan or a boolean value, expression'
-                                 '"%s" has units %s instead') % (ref, unit))
+                                 '"%s" has units %s instead') % (ref, dims))
```
Same issue as the fixme above?

thesamovar commented on brian2/groups/neurongroup.py:
```diff
@@ -603,7 +602,7 @@ def __setattr__(self, key, value):
 
             eq = self.equations[key]
-            if eq.unit != linked_var.unit:
+            if eq.dim != linked_var.dim:
```
Here we're using ``!=`` instead of ``is not`` so I guess this can work then?

thesamovar commented on brian2/parsing/expressions.py:
```diff
@@ -181,7 +182,7 @@ def _get_value_from_expression(expr, variables):
-def parse_expression_unit(expr, variables):
+def parse_expression_dimensions(expr, variables):
     '''
     Returns the unit value of an expression, and checks its validity
```
Again, change docstring to say physical dimensions?

thesamovar commented on brian2/tests/test_equations.py:
```diff
@@ -124,7 +124,7 @@ def test_parse_equations():
     eqs = parse_string_equations('dv/dt = -v / tau : 1')
     assert len(eqs.keys()) == 1 and 'v' in eqs and eqs['v'].type == DIFFERENTIAL_EQUATION
-    assert get_dimensions(eqs['v'].unit) == DIMENSIONLESS
+    assert eqs['v'].dim is DIMENSIONLESS
```
is check again

thesamovar commented on brian2/units/fundamentalunits.py:
```diff
@@ -191,7 +190,7 @@ def wrap_function_dimensionless(func):
     Returns a new function that wraps the given function `func` so that it
     raises a DimensionMismatchError if the function is called on a quantity
-    with dimensions (excluding dimensionless quantitities). Quantities are
+    with dimensions (excluding dimensionless quantities). Quantities are
```
Maybe physical dimensions here too?

thesamovar commented on brian2/units/fundamentalunits.py:
```diff
@@ -460,7 +458,11 @@ def __ipow__(self, value):
     #### COMPARISON ####
     def __eq__(self, value):
-        return np.allclose(self._dims, value._dims)
+        try:
```
If we change the is checks to use an ``==`` check instead, we could first check if ``self._dims is value._dims`` and potentially get most of the speed still.

thesamovar commented on brian2/units/fundamentalunits.py:
```diff
@@ -2237,69 +2240,50 @@ def unregister_unit(u):
-    a dimensionless quantity
-    '''
-    if x is 1 or x is DIMENSIONLESS or have_same_dimensions(x, 1):
+    if d is 1 or d is DIMENSIONLESS:
```
Same issue?

mstimberg commented:
I adapted the docstrings and changed "dimensions" to "physical dimensions" everywhere (I hope!). I agree that should make things clearer (well, "physical dimensions" could also mean something like width x height I guess, but that should be clear enough from the context...). I am not that worried about the `ndim` attribute, we only have in one place (`DynamicArrayVariable`) and either way it should not be directly used by user code.

I am not sure about the `Dimension.time` thing. It certainly would look cleaner than `second.dim`, but then we'd also need `Dimension.conductance`, etc. (given that we don't want anyone to write``Dimension.length**-2 * Dimension.mass**-1 * Dimension.time**3 * Dimension.electrical_current**2``...). For now, I tried to make the functions that expect dimensions to be "forgiving", i.e. you could actually write `dimensions=second` instead of `dimensions=second.dim`. I guess it is better to not add new layer of names; as we said earlier, users don't care that much about the distinction between units and dimensions.

About the check with `is` vs. `==` -- actually this is designed to be ok. No code should ever create new `Dimension` objects explicitly, but always pass through `get_or_create_dimensions` which will return a `Dimension` object from a cache if it already exists. So stuff like `second.dim * second.dim is second.dim**2` *will* evaluate to ``True``.

I improved one error message (the previous FIXME), so if the tests pass I am going to merge this.

thesamovar commented:
Great!
